addons/udes_report/models/stock_export.py

```diff
@@ -62,79 +62,99 @@ def run_stock_file_export(self, context=None, send_file_via='user'):
 
+    def _generate_stock_file_data(self):
+        '''
+        Generate data for the stock file.
+
+        This will contain 2 sheets that may then be exported to any format
+        (currently only excel)
+        Sheets:
+            1) part number, location, package ID and quantity;
+            2) part number, number of packages and total quantity.
+        '''
         Quant = self.env['stock.quant']
         Product = self.env['product.product']
         Location = self.env['stock.location']
-        all_quants = Quant.search([('location_id', 'child_of', locations.ids)])
-        by_prod = lambda x: x.product_id.id
-        by_location = lambda x: x.location_id.id
-        by_package_name = lambda x: x.package_id.name or ''
 
-        get_quants_by_loc = lambda quant_union: {
-            Location.browse(loc): Quant.union(*_quants)
-            for loc, _quants in groupby(quant_union.sorted(key=by_location),
-                                        key=by_location)}
+        def by_prod(record):
+            return record.product_id.id
+
+        def by_location(record):
+            return record.location_id.id
+
+        def by_package_name(record):
+            return record.package_id.name or ''
+
+        def by_lot_name(record):
+            return record.lot_id.name or ''
+
+        def get_quants_by_loc(quant_union):
+            return {Location.browse(loc): Quant.union(*_quants)
+                    for loc, _quants in groupby(quant_union.sorted(key=by_location),
+                                                key=by_location)}
+
+        locations = self.included_locations - self.excluded_locations
+        if not locations:
+            raise UserError(_("The specified list of Stock Locations is empty."))
+
+        all_quants = Quant.search([('location_id', 'child_of', locations.ids)])
 
         quants_by_prod_by_loc = OrderedDict()
         prod_summary = defaultdict(lambda: {'n_pkgs': 0, 'qty': 0})
 
-        for prod, _quants in groupby(all_quants.sorted(key=by_prod),
-                                     key=by_prod):
+        for prod, _quants in groupby(all_quants.sorted(key=by_prod), key=by_prod):
             quant_prods = Product.browse(prod)
             quants_u = Quant.union(*_quants)
             quants_by_prod_by_loc[quant_prods] = get_quants_by_loc(quants_u)
 
-        def get_prod_qty(quants):
-            return sum(Quant.union(*quants).mapped('quantity'))
-
-        _logger.info(_('Creating Excel file'))
-        file_name = 'warehouse_stock_{TIMESTAMP}.xls'.format(
-            TIMESTAMP=datetime.now())
-        wb = xlwt.Workbook()
-
-        # 1/2) 'Stock File': one entry for each (prod, package) pair
+        # 1/2) 'Stock File': one entry for each prod, package and lot (if applicable)
+        stock_file_data = {"name": 'Stock File'}
+        # Determine if lots/serials should be shown on the report
+        show_tracking = bool(Product.search_count([("tracking", "!=", "none")]))
```

Comment:
Nit: Can also add a ```limit=1``` and ```order=id desc``` as you just need to know if any results are returned, would probably only give the most minimal of performance increases though. :)

Reply:
Actually as it's ```search_count``` then ```order``` probably isn't relevant. Feel free to resolve anyway. :)

Reply:
Changed to 'Product.search([("tracking", "!=", "none")], order="id", limit=1)' as discussed on skype